### ISSUE TITLE
Bogus Error in Measure Definition

### ISSUE THREAD

**Issue by ranjitjhala** [ucsd-progsys/liquidhaskell#427]
See this:

http://goto.ucsd.edu:8090/index.html#?demo=permalink%2F1439294454_1547.hs

aka: `tests/todo/doubleError.hs`

Right now, the error you get is:

```
 doub.hs:11:5: Error: Bad Type Specification
 Blank.ok :: {v : Double | myprop v}
     Sort Error in Refinement: {v : real | myprop v}
     BExp myprop([v]) with non-propositional type GHC.Types.Bool
```

This is because the user has written `Bool` instead of `Prop` -- and how are they to know! So fine,
if I change `Bool` to `Prop` that error goes away and it says "unsafe" (at the 6.0). I was mystified, how is this super complex thing promoted to a measure? And what measure is it promoted to? I recalled that
it should be enough to just say 

```
{-@ measure myprop @-}
```

and when I do that I get:

```
 doub.hs:3:13: Error: Cannot promote Haskell function myprop to logic
     Measure Functions should have a case at top level
```

which is at least closer to what I'd expect.

So whats going on?

1. Why does adding/removing the type signature in the measure cause different behavior?
2. Can we just flag an error if the measure has a type signature because it should not be there?

Anyways, we should fix this, @nikivazou.

**Comment by nikivazou** [comment]
So, the function `myprop` cannot be promoted to a measure because of its structure and this is why when you write 
```
{-@ measure myprop @-}
```
you get the (expected) Error.

The `measure` keyword is used in two cases

- `{-@ measure fromHaskell @-}` promotes the Haskell function `fromHaskell` into logic.
- `{-@ measure uninterpretedMeasure :: t @-}` _defines_ an uninterpreted measure `uninterpretedMeasure`. This behavior was there from the very early stages on liquidHaskell. 

We need both behaviors, but we could change the syntax. Maybe we could have
- `{-@ lift measure fromHaskell @-}` to promote the Haskell function `fromHaskell` into logic.
- `{-@ define measure uninterpretedMeasure :: t @-}` to  _define_ the uninterpreted measure `uninterpretedMeasure`.

**Comment by ranjitjhala** [comment]
Do we really need the extra keyword? Why not just:

1 omit the type when there is a concrete measure Implementation; 

2 supply the type for abstract measures

If you supply the type AND have an implementation that should be an error.

?

**Comment by gridaphobe** [comment]
What if I want to use an abstract measure to talk about a Haskell function that cannot be lifted? For example, in Data.Text I used

    measure ord :: Char -> Int
    ord :: c:Char -> {v:Int | v = ord c}

to treat ord as an uninterpreted function, which was crucial for memory safety.

**Comment by ranjitjhala** [comment]
Well in this case you could have as well named the measure something else -- no need to reuse the same name as the haskell function.

**Comment by gridaphobe** [comment]
Sure, but I think reusing the Haskell name is most appropriate because it signifies unambiguously what the measure is supposed to denote.

**Comment by ranjitjhala** [comment]
I see your point but there's also a bit misleading or confusing because a user may think arbitrary functions are magically converted into the logic which is, of course not the case...

**Comment by nikivazou** [comment]
I agree with Eric. Also, as Ranjit said it is confusing to change the
semantics of the keyword 'measure' according to if there is a type
signature or not.

Let's create two different one-word keywords one for measure definition and
one for lifting.

That being said my question is what exactly is a measure? Did you introduce
it at Liquid types to denote the recursive logical functions? If so, the
second measure case does not define a measure, but just an uninterpreted
function.

**Comment by ranjitjhala** [comment]
Good point: the second case is not really a measure, so maybe we should
just call it something else (and reserve measure for concrete things like
len, etc. which actually have implementations.)

On Wed, Aug 12, 2015 at 4:02 PM, Niki Vazou <notifications@github.com>
wrote:

> I agree with Eric. Also, as Ranjit said it is confusing to change the
> semantics of the keyword 'measure' according to if there is a type
> signature or not.
>
> Let's create two different one-word keywords one for measure definition and
> one for lifting.
>
> That being said my question is what exactly is a measure? Did you introduce
> it at Liquid types to denote the recursive logical functions? If so, the
> second measure case does not define a measure, but just an uninterpreted
> function.
> On Aug 11, 2015 12:46 PM, "Ranjit Jhala" <notifications@github.com> wrote:
>
> > I see your point but there's also a bit misleading or confusing because a
> > user may think arbitrary functions are magically converted into the logic
> > which is, of course not the case...
> >
> >
> > > On Aug 11, 2015, at 10:03 PM, Eric Seidel <notifications@github.com>
> > wrote:
> > >
> > > Sure, but I think reusing the Haskell name is most appropriate because
> > it signifies unambiguously what the measure is supposed to denote.
> > >
> > > —
> > > Reply to this email directly or view it on GitHub.
> > >
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/ucsd-progsys/liquidhaskell/issues/427#issuecomment-129963329
> >
>
> > .
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/issues/427#issuecomment-130251693>
> .
>



-- 
Ranjit.

**Comment by nikivazou** [comment]
Maybe just call the second case `function` (as uninterpreted is a scary word) or `definition`?

**Comment by gridaphobe** [comment]
Instead of adding two new keywords, why not just one: `abstract`? We would partition measures more explicitly between those that have bodies (regular measures) and those that only have a signature (abstract measures). A few examples:

```
{-@ abstract measure foo :: Bar -> Int @-}
-- due to the `abstract` modifier, `foo` CANNOT have a definition

{-@ measure length @-}
length = ...
-- as before

{-@ measure goo :: Bar -> Int @-}
-- ERROR: measures must have a body or be lifted
```

**Comment by ranjitjhala** [comment]
Yes, this is probably the best thing IMO.

On Wed, Aug 12, 2015 at 9:04 PM, Eric Seidel <notifications@github.com>
wrote:

> Instead of adding two new keywords, why not just one: abstract? We would
> partition measures more explicitly between those that have bodies (regular
> measures) and those that only have a signature (abstract measures). A few
> examples:
>
> {-@ abstract measure foo :: Bar -> Int @-}
> -- due to the `abstract` modifier, `foo` CANNOT have a definition
>
> {-@ measure length @-}
> length = ...
> -- as before
>
> {-@ measure goo :: Bar -> Int @-}
> -- ERROR: measures must have a body or be lifted
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/issues/427#issuecomment-130344272>
> .
>



-- 
Ranjit.

**Comment by nikivazou** [comment]
But we said that "abstract measures" are not measures. They are just
uninterpreted functions

On Wed, Aug 12, 2015 at 8:41 AM, Ranjit Jhala <notifications@github.com>
wrote:

> Yes, this is probably the best thing IMO.
>
> On Wed, Aug 12, 2015 at 9:04 PM, Eric Seidel <notifications@github.com>
> wrote:
>
> > Instead of adding two new keywords, why not just one: abstract? We would
> > partition measures more explicitly between those that have bodies
> (regular
> > measures) and those that only have a signature (abstract measures). A few
> > examples:
> >
> > {-@ abstract measure foo :: Bar -> Int @-}
> > -- due to the `abstract` modifier, `foo` CANNOT have a definition
> >
> > {-@ measure length @-}
> > length = ...
> > -- as before
> >
> > {-@ measure goo :: Bar -> Int @-}
> > -- ERROR: measures must have a body or be lifted
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <
> https://github.com/ucsd-progsys/liquidhaskell/issues/427#issuecomment-130344272
> >
> > .
> >
>
>
>
> --
> Ranjit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/issues/427#issuecomment-130347259>
> .
>



-- 
Niki Vazou

**Comment by gridaphobe** [comment]
Measures are also "just" uninterpreted functions. We just happen to add axioms about them whenever we see a DataCon

**Comment by nikivazou** [comment]
I am not sure, but I think measures are uninterpreted functions PLUS the
axioms or recursive definitions. Ie, arbitrary uninterpreted functions are
NOT measures

**Comment by gridaphobe** [comment]
In my mind, measures are just a (very) convenient syntax for axiomatizing the behavior of uninterpreted functions. What's the difference between

```
data L a = N | C a (L a)

measure len :: L a -> Int
len N = 0
len (C x xs) = 1 + len xs
```

and

```
data L a where
  N :: {v:L a | len v = 0}
  C :: x:a -> xs:L a -> {v:L a | len v = 1 + len xs}

abstract measure len :: L a -> Int
```

We don't support the latter syntax at the moment IIRC, but there's no reason we couldn't.

**Comment by ranjitjhala** [comment]
Agree with Eric.

Niki: just mentally replace measure with "property" or "thing". Now if you say

"Abstract thing" it just means a thing without any definitions or axioms (much like an "abstract class" in Java)

**Comment by nikivazou** [comment]
Well, that goes back to my initial question: who introduced the term measure and what does it mean?

I have seen somewhere, I think in Colin's work, that measures where introduced by liquid types to measure the size of structures. But I never found the definition in the liquid types paper. If that is correct, then uninterpreted functions technically are not measures, but just uninterpreted functions. I know I am pedantic here. 

I prefer "function" instead of "abstract measure". Why do we required somebody to understand what a measure is if she is only to use it as an uninterpreted "function"? I mean the term function is much more well understood and simple.

**Comment by ranjitjhala** [comment]
This discussion has become too abstract to conduct over email ;) 
let's defer to next week when I'm back? 

Ps: it's introduced in the pldi09 paper.
